### ISSUE TITLE
Potential Bug in jax.grad() and jax.jacrev()

### ISSUE THREAD

**Issue by LeoXinhaoLee** [google/jax#19727]
### Description

Hi, when using jax.vmap to decorate jax.grad() or using jax.jacrev to calculate the Jacobian of a loss function which includes LayerNorm, it seems like the first returned gradient is 0, which is not expected. Could you please help me with this? Thank you.

Two examples that should be equivalent and both had the described problems are as follows:

(1)
```
import flax.linen as nn

rng = jax.random.PRNGKey(0)

X = jax.random.normal(rng, (2, 8))             # the batch dim is 2, which has the problem
target = jax.random.normal(rng, (2, 2))     # the batch dim is 2, which has the problem

A = jax.random.normal(rng, (4, 2))
W1_init = jax.random.normal(rng, (8, 16))
W2_init = jax.random.normal(rng, (16, 4))

LN = nn.LayerNorm(use_bias=False)
LN_params = LN.init(rng, jnp.ones((1, 2)))['params']

def loss(W1, W2, LN_params, A, X, target):
    XWA = X @ W1 @ W2 @ A
    LN_out = LN.apply({"params": LN_params}, XWA)
    l = 0.5 * ((LN_out - target) ** 2).sum()
    return l

grad_fn = jax.value_and_grad(loss, argnums=[0,1])

l1, grads = vmap(partial(grad_fn, W1_init, W2_init, LN_params, A))(X.reshape(-1,1,8), target.reshape(-1,1,2))

grad_w1, grad_w2 = grads[0], grads[1]

print(grad_w1[0])  # all 0 !!!
print(grad_w2[0])  # all 0 !!!
```

(2)
```
import flax.linen as nn

rng = jax.random.PRNGKey(0)

X = jax.random.normal(rng, (2, 8))
target = jax.random.normal(rng, (2, 2))

A = jax.random.normal(rng, (4, 2))
W1_init = jax.random.normal(rng, (8, 16))
W2_init = jax.random.normal(rng, (16, 4))

LN = nn.LayerNorm(use_bias=False)
LN_params = LN.init(rng, jnp.ones((1, 2)))['params']

def loss(W1, W2, LN_params, A, X, target):
    XWA = X @ W1 @ W2 @ A
    LN_out = LN.apply({"params": LN_params}, XWA)
    l = 0.5 * ((LN_out - target) ** 2).sum(axis=-1)
    return l, l

grad_fn = jax.jacrev(loss, argnums=[0,1], has_aux=True)

grads, l1 = grad_fn(W1_init, W2_init, LN_params, A, X, target)

grad_w1, grad_w2 = grads[0], grads[1]

print(grad_w1[0])  # all 0 !!!
print(grad_w2[0])  # all 0 !!!
```

In both (1) and (2), grad_w2's shape is [2, 16, 4], but both grad_w1[0] and grad_w2[0] is **all 0s**, which is not expected. If in these examples I set the first dimension of `X` and `target` **to be 1 or larger than 2, this problem will not appear**.

Or if there's **no LayerNorm** in the `loss` function, the problem will not appear.

Thank you very much for your time and help! @jakevdp 

### What jax/jaxlib version are you using?

flax==0.6.8; jax==0.4.13; jaxlib==0.4.13+cuda12.cudnn89

### Which accelerator(s) are you using?

GPU

### Additional system info?

Ubuntu 20.04

### NVIDIA GPU info

<img width="896" alt="image" src="https://github.com/google/jax/assets/62500792/2dd5ac93-21e1-48b4-b026-590a540e4348">

**Comment by mattjj** [comment]
Thanks for the question.

Can you try updating to a newer version? I am unable to reproduce at HEAD. I think there was a bug in the differentiation rule for `x ** 2` in a version, which may have been 0.4.13. But pypi is on 0.4.24 now, so we've had quite a few releases since then

**Comment by LeoXinhaoLee** [comment]
Thank you for your kind response. It seems like in my above examples, grad_w1[0] is mistakenly automatically added to grad_w1[1]. 

But anyway, I tried jax==0.4.20 and flax==0.7.5, still have the same problem. The latest version seems to be fine, but cannot work on my GPU due to cublas version.

Could you please try my examples on some older jax version and see if it's fixable? I guess it would be a challenge for me to move my entire codebase to a new jax version. If it's not convenient for you, it's also fine. Thank you so much!

**Comment by mattjj** [comment]
The bug fix I had in mind was #17996, but that was included in jax==0.4.20. So that's not it.

Are you sure the correct gradient isn't zero? I put this at the bottom of the file and ran it a few times to plot the function in random directions, and it sure looks flat at 0 (i.e. at `W1_init` and `W2_init`):

```python
import numpy.random as npr
from matplotlib import pyplot as plt
v_W1 = npr.randn(*W1_init.shape)
v_W2 = npr.randn(*W2_init.shape)
ts = jnp.linspace(-1, 1, 1000)
vals = jax.vmap(lambda t: loss(W1_init + t * v_W1, W2_init + t * v_W2,
                               LN_params, A, X, target))(ts)
plt.plot(ts, vals)
plt.savefig('19727.png')
```

**Comment by LeoXinhaoLee** [comment]
Thank you for your kind response! I appended your code to mine and get this plot, is this also what you see?

```
import numpy as np
from matplotlib import pyplot as plt
import flax.linen as nn

rng = jax.random.PRNGKey(0)

X = jax.random.normal(rng, (2, 8))             # the batch dim is 2, which has the problem
target = jax.random.normal(rng, (2, 2))     # the batch dim is 2, which has the problem

A = jax.random.normal(rng, (4, 2))
W1_init = jax.random.normal(rng, (8, 16))
W2_init = jax.random.normal(rng, (16, 4))

LN = nn.LayerNorm(use_bias=False)
LN_params = LN.init(rng, jnp.ones((1, 2)))['params']

def loss(W1, W2, LN_params, A, X, target):
    XWA = X @ W1 @ W2 @ A
    LN_out = LN.apply({"params": LN_params}, XWA)
    l = 0.5 * ((LN_out - target) ** 2).sum()
    return l

grad_fn = jax.value_and_grad(loss, argnums=[0,1])

l1, grads = vmap(partial(grad_fn, W1_init, W2_init, LN_params, A))(X.reshape(-1,1,8), target.reshape(-1,1,2))

grad_w1, grad_w2 = grads[0], grads[1]

print(grad_w1[0])  # all 0 !!!
print(grad_w2[0])  # all 0 !!!

v_W1 = np.random.randn(*W1_init.shape)
v_W2 = np.random.randn(*W2_init.shape)
ts = jnp.linspace(-1, 1, 1000)
vals = jax.vmap(lambda t: loss(W1_init + t * v_W1, W2_init + t * v_W2,
                               LN_params, A, X, target))(ts)
plt.plot(ts, vals)
plt.savefig('19727.png')
```
<img width="554" alt="image" src="https://github.com/google/jax/assets/62500792/2827a148-2c98-481c-ab58-3136bef53a8f">

**Comment by mattjj** [comment]
Yes, qualitatively, though every time you run it you'll get a different random direction. It looks piecewise constant so I'd expect a derivative of 0. (I didn't reason through the code as to why it's piecewise constant, but this evaluation has no autodiff in it.)

**Comment by LeoXinhaoLee** [comment]
I see. It seems like I happen to run into a test case that gives me a very flat loss at this specific point, and the result of auto diff is not unexpected. Thank you very much for your time and help!

**Comment by mattjj** [comment]
No problem! Please reopen or file new issues as things come up.
